solutions/src/main/java/ru/sberbank/school/task10/ScalableThreadPool.java

```diff
@@ -0,0 +1,139 @@
+package ru.sberbank.school.task10;
+
+import lombok.NonNull;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+
+public class ScalableThreadPool implements ThreadPool {
+    private final Queue<Runnable> taskQueue;
+
+    private final int min;
+    private final int max;
+    private volatile int workingThreads = 0;
+    private final List<ThreadWorker> threads;
+    private final List<ThreadWorker> excessiveThreads;
+    private volatile int freeId;
+
+    public ScalableThreadPool(int min, int max) {
+        if (min <= 0 || max <= 0 || min > max) {
+            throw new IllegalArgumentException();
+        }
+        this.taskQueue = new LinkedList<>();
+        this.min = min;
+        this.max = max;
+        this.freeId = min + 1;
+        this.threads = new ArrayList<>(min);
+        this.excessiveThreads = new ArrayList<>(max-min);
+    }
+
+    @Override
+    public void start() {
+        if (!threads.isEmpty()) {
+            throw new IllegalStateException("Пул уже запущен!");
+        }
+        synchronized (threads) {
+            for (int i = 0; i < min; i++) {
+                ThreadWorker threadWorker = new ThreadWorker(null);
+                threads.add(threadWorker);
+                threadWorker.start();
+            }
+            workingThreads = min;
+        }
+    }
+
+    @Override
+    public synchronized void stopNow() {
+        taskQueue.clear();
+        deleteExcessiveThreads();
+        for (ThreadWorker t : threads) {
+            t.interrupt();
+        }
+        threads.clear();
+        excessiveThreads.clear();
+        resetFreeId();
+        workingThreads = 0;
+    }
+
+    @Override
+    public synchronized void execute(@NonNull Runnable runnable) {
+        if (threads.isEmpty()) {
+            throw new IllegalStateException("Пул еще не запущен!");
+        }
+        if (taskQueue.isEmpty()) {
+            if (workingThreads > min) {
+                deleteExcessiveThreads();
+            }
+        } else if (workingThreads <= min && workingThreads <= max) {
+            ThreadWorker threadWorker = new ThreadWorker(runnable);
+            excessiveThreads.add(threadWorker);
+            threadWorker.start();
+            workingThreads++;
+            return;
+        }
+
+        synchronized (taskQueue) {
+            taskQueue.add(runnable);
+            taskQueue.notify();
+        }
+    }
+
+    private void deleteExcessiveThreads() {
+        for (ThreadWorker excessiveThread : excessiveThreads) {
+            excessiveThread.interrupt();
+        }
+        excessiveThreads.clear();
+        workingThreads = min;
+        resetFreeId();
+    }
+
+    private void resetFreeId() {
+        freeId = min + 1;
+    }
+
+    @Override
```

Comment:
Overload methods should not be split. Previous overloaded method located at line &#39;63&#39;.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a598559860>*